jupyterhub/objects.py

```diff
@@ -26,10 +27,22 @@ class Server(HasTraits):
     orm_server = Instance(orm.Server, allow_none=True)
 
     ip = Unicode()
+    connect_ip = Unicode()
     proto = Unicode('http')
     port = Integer()
     base_url = Unicode('/')
     cookie_name = Unicode('')
+    
+    @property
+    def _connect_ip(self):
+        """Property for connect_ip"""
+        if self.connect_ip:
+            return self.connect_ip
+        elif self.ip in {'', '0.0.0.0'}:
+            # if listening on all interfaces, default to hostname
```

Comment:
Binding to socketname has very different (and IMO unexpected) behavior from binding to 0.0.0.0.

```bash
python3 -m http.server --bind 0.0.0.0 8000
```

curl localhost:8000 works, curl courage:8000 works (courage is my hostname).

```
python3 -m http.server --bind courage 8000
```

curl localhost:8000 does *not* work, and curl courage:8000 does work. If we do this substitution by default you can never listen on localhost and public interfaces, even though you specified 0.0.0.0 which has that behavior - unintuitive and confusing.

If people want it to only listen on a public interface, they should be asked to explicitly specify that IMO.